### PR TITLE
[app_dart] Task statuses can be updated without task keys

### PR DESCRIPTION
**Pull request by CaseyHillers**
Migrating the DeviceLab to LUCI requires updating Cocoon to work with what LUCI knows. LUCI runs will not know the task key they are running. Instead, they know the git sha, git branch, and task name. This updates the request handler to process those to find corresponding task.

# Issue

https://github.com/flutter/flutter/issues/66191

# Tests

Added a unit test for this use case

### CODE REVIEW

File: app_dart/test/request_handlers/update_task_status_test.dart [hunk 1]
```diff
@@ -120,5 +120,31 @@ void main() {
       expect(task.status, 'Failed');
       expect(task.attempts, 1);
     });
+
+    test('non-task key requests can update tasks', () async {
```
Comment (keyonghan) [inline app_dart/test/request_handlers/update_task_status_test.dart:124]: one more test for cases where no task is found?

File: app_dart/lib/src/request_handlers/update_task_status.dart [hunk 2]
```diff
@@ -21,6 +21,20 @@ import '../request_handling/body.dart';
 import '../request_handling/exceptions.dart';
 import '../service/datastore.dart';
 
+/// Endpoint for task runners to update Cocoon with test run information.
+///
+/// This handlers requires (1) task identifier and (2) task status information.
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/update_task_status.dart:26]: nit: handler

File: app_dart/lib/src/request_handlers/update_task_status.dart [hunk 3]
```diff
@@ -111,6 +117,58 @@ class UpdateTaskStatus extends ApiRequestHandler<UpdateTaskStatusResponse> {
...
+
+  /// Retrieve [Task] from [DatastoreService] when given [gitShaParam], [gitBranchParam], and [taskNameParam].
+  ///
+  /// This is used when the DeviceLab test runner is uploading results to Cocoon for runs on LUCI.
+  /// LUCI does not know the [Key] assigned to task when scheduling the build, but Cocoon can
+  /// lookup the task based on these key values.
+  ///
+  /// To lookup the value, we construct the ancestor key, which corresponds to the [Commit].
+  /// Then we query the tasks with that ancestor key and filter for the one that has the corresponding name.
+  Future<Task> _getTaskFromNamedParams(DatastoreService datastore) async {
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/update_task_status.dart:152]: add checkRequiredParameters for [gitShaParam], [gitBranchParam], [taskNameParam] when the task is from luci.

File: app_dart/test/request_handlers/update_task_status_test.dart (hunk 1 above)
Comment (CaseyHillers) [reply app_dart/test/request_handlers/update_task_status_test.dart]: Good catch, added.

File: app_dart/lib/src/request_handlers/update_task_status.dart (hunk 3 above)
Comment (CaseyHillers) [reply app_dart/lib/src/request_handlers/update_task_status.dart:157]: Added check for this when from luci, otherwise verify taskKeyParam was passed.

**Review by keyonghan — APPROVED** [review]
LGTM

**Pull request merged**